### DISCUSSION TITLE
Development of client and server for device-to-device connection

### DISCUSSION THREAD

**Discussion by Lafatalite** [pyri-project/pyri-core#14] [Q&A]
Hello, I am currently developing a project similar to robotics. At present, I can successfully start the node using PyriServiceNodeSetup and complete communication through a simple robotraconteur client. However, when I start through pyricore, it reports an error and I cannot add this device. The corresponding webui also cannot 
discover this device and thus cannot connect. Am I missing something?

[demo.tar.gz](https://github.com/user-attachments/files/15612756/demo.tar.gz)
<img width="700" alt="Snipaste_2024-06-06_17-13-27" src="https://github.com/pyri-project/pyri-core/assets/35282465/21405ae1-26f3-4eda-92d0-0d3418ae6e1b">
<img width="732" alt="Snipaste_2024-06-06_17-14-43" src="https://github.com/pyri-project/pyri-core/assets/35282465/9fe6008c-293e-488d-811e-8d11b7b60ad8">

**Comment by johnwason** [comment]
A quick review looks ok.

There is the line:

```
with PyriServiceNodeSetup("tech.pyri.robot.demostrator", 58654
```

Is something else using port 58654? Do the pyri node logs show a conflict?

**Reply by Lafatalite** [reply]
I also tried changing the port but it still didn't work. If it can be added to the default device, can I use `self.core.device_manager.get_device_subscription("robot_demonstrator")` to get its object for RPC access?

**Reply by johnwason** [reply]
You don't need to bother with the launcher or default device if you just want to communicate with the service. If the service implements com.robotraconteur.device.Device you can add it manually using the Devices tab.

**Reply by Lafatalite** [reply]
I have a corresponding browser to communicate with it. I want to call the functions of the `demonstrator` in this corresponding browser. Currently, I use `self.core.device_manager.get_device_subscription("robot_demonstrator")` to connect to the device, but it returns a NoneType because it caught an error.

<img width="774" alt="Snipaste_2024-06-07_09-18-41" src="https://github.com/pyri-project/pyri-core/assets/35282465/b6b92987-af21-4cd1-a8bb-a93ccff19a9b">

<img width="520" alt="Snipaste_2024-06-07_09-19-44" src="https://github.com/pyri-project/pyri-core/assets/35282465/350727e9-3bb0-4bd4-a5f5-de913f5ef841">

**Reply by johnwason** [reply]
I would suggest using AsyncGetDefaultClient instead of GetDefaultClient, as that will wait until the client has connected. See https://github.com/pyri-project/pyri-robotics-browser/blob/549fecff9b8b291be2a2f499a9f22eecf5fdc069/src/pyri/robotics_browser/components/jog_component.py#L223 for an example.

This looks like robot_demostrator is not added as a device. Do you see robot_demostrator as connected on the "Devices" tab?

**Reply by johnwason** [reply]
Did you have other issues/discussions I need to look at? I may have lost track of them...

**Reply by Lafatalite** [reply]
Oh, I've already resolved that issue. There were no problems with the browser call, and I was able to get the client using `AsyncGetDefaultClient`
Now, I have a new question. I want to develop a node to be deployed on the robot, which drives the robotic arm by calling its API. However, the `pyri-core` is running on another server. Can I develop a `RobotRaconteur` node that connects via TCP? Or how do you usually handle this?

**Reply by johnwason** [reply]
If you want to control the robot using the Blockly or Python scripts, you would use sandbox functions. Those run in the pyri-sandbox node which is a normal python program that does not run in the browser. If you want to control the robot from something like a user interface in the browser, I would recommend using an intermediate node similar to how the robotics-jog works.

**Reply by Lafatalite** [reply]
I think my description had an issue. I want to develop a device node to be deployed separately on the robot. How can I develop this device node? Should I use `PyriServiceNodeSetup?` However, the environment on the robot might not have the pyri environment deployed. If I use this function, won't it depend on components like `pyri-core` during startup? Or can I use the `RobotRaconteur` framework to develop a node?
